internal/pkg/flink/app/application.go

```diff
@@ -1,62 +1,165 @@
+	app := Application{
+		history:                     historyStore,
+		store:                       dataStore,
+		resultFetcher:               resultFetcher,
+		appController:               appController,
+		inputController:             inputController,
+		statementController:         statementController,
+		interactiveOutputController: interactiveOutputController,
+		basicOutputController:       basicOutputController,
+		tokenRefreshFunc:            synchronizedTokenRefresh(tokenRefreshFunc),
+	}
+	components.PrintWelcomeHeader()
+	app.readEvalPrintLoop()
+}
 
-	// Event handlers
-	app.SetInputCapture(tableController.AppInputCapture)
+func getStdin() *term.State {
+	state, err := term.GetState(int(os.Stdin.Fd()))
+	if err != nil {
+		log.CliLogger.Warnf("Couldn't get stdin state with term.GetState. Error: %v\n", err)
+		return nil
+	}
+	return state
+}
 
-	interactiveOutput := components.InteractiveOutput(table, shortcuts)
-	rootLayout := components.RootLayout(interactiveOutput)
+func getConsoleParser() prompt.ConsoleParser {
+	consoleParser := prompt.NewStandardInputParser()
+	err := consoleParser.Setup()
+	if err != nil {
+		log.CliLogger.Warnf("Couldn't setup console parser. Error: %v\n", err)
+	}
+	return consoleParser
+}
+
+func tearDownConsoleParser(consoleParser prompt.ConsoleParser) {
+	err := consoleParser.TearDown()
+	if err != nil {
+		log.CliLogger.Warnf("Couldn't tear down console parser. Error: %v\n", err)
+	}
+}
 
-	// We start tview and then suspend it immediately so we initialize all components
-	app.SetAfterDrawFunc(func(screen tcell.Screen) {
-		if !screen.HasPendingEvent() {
-			once.Do(func() {
-				go appController.SuspendOutputMode(inputController.RunInteractiveInput)
-			})
+func restoreStdin(state *term.State) {
+	if state != nil {
+		_ = term.Restore(int(os.Stdin.Fd()), state)
+	}
+}
+
+func (a *Application) readEvalPrintLoop() {
+	for a.isAuthenticated() {
+		userInput := a.inputController.GetUserInput()
+		if a.inputController.IsSpecialInput(userInput) {
+			continue
 		}
-	})
+		a.history.Append(userInput)
+
+		executedStatement, err := a.statementController.ExecuteStatement(userInput)
+		if err != nil {
+			continue
+		}
+
+		executedStatementWithResults, err := a.fetchInitialResults(*executedStatement)
+		if err != nil {
+			continue
+		}
+		a.resultFetcher.Init(*executedStatementWithResults)
+		a.getOutputController(*executedStatementWithResults).VisualizeResults()
+	}
+}
 
-	// Start the application.
-	if err := appController.StartTView(rootLayout); err != nil {
-		panic(err)
+func (a *Application) isAuthenticated() bool {
+	if authErr := a.tokenRefreshFunc(); authErr != nil {
+		utils.OutputErrf("Error: %v\n", authErr)
+		a.appController.ExitApplication()
+		return false
 	}
+	return true
+}
+
+func (a *Application) fetchInitialResults(executedStatement types.ProcessedStatement) (*types.ProcessedStatement, *types.StatementError) {
```

Comment:
I like that we differentiate both 👍 

Maybe `fetchFirstPageResults` would be a better name? what do you think

Reply:
yeah I think that sounds good! will rename 👍

Reply:
actually just realized that function might be a bit unnecessary, since it's only used in that one spot.